src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,465 @@
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <base/getFQDNOrHostName.h>
+#include <base/logger_useful.h>
+
+#include <Common/Exception.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/setThreadName.h>
+
+#include <Core/ServerUUID.h>
+
+#include <Interpreters/Cluster.h>
+#include <Interpreters/ClusterDiscovery.h>
+#include <Interpreters/Context.h>
+
+#include <Poco/Exception.h>
+#include <Poco/JSON/JSON.h>
+#include <Poco/JSON/Object.h>
+#include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+fs::path getShardsListPath(const String & zk_root)
+{
+    return fs::path(zk_root + "/shards");
+}
+
+}
+
+/*
+ * Holds boolean flags for fixed set of keys.
+ * Flags can be concurrently set from different threads, and consumer can wait for it.
+ */
+template <typename T>
+class ClusterDiscovery::ConcurrentFlags
+{
+public:
+    template <typename It>
+    ConcurrentFlags(It begin, It end)
+    {
+        for (auto it = begin; it != end; ++it)
+            flags.emplace(*it, false);
+    }
+
+    void set(const T & key)
+    {
+        setFlag(key, true);
+        cv.notify_one();
+    }
+
+    /// `need_update` expected to be value from `flags` corresponding to some key
+    void set(std::atomic_bool & need_update)
+    {
+        any_need_update = true;
+        need_update = true;
```

Comment:
Looks like we need to change an order?
And remove `need_update` argument from here - it looks strange.

Reply:
removed this method overload at all